Source/WebCore/css/CSSValue.cpp

```diff
@@ -249,264 +234,33 @@ bool CSSValue::isCSSLocalURL(StringView relativeURL)
-    case ConicGradientClass:
-        return downcast<CSSConicGradientValue>(*this).customCSSText();
-    case CrossfadeClass:
-        return downcast<CSSCrossfadeValue>(*this).customCSSText();
-    case ImageClass:
-        return downcast<CSSImageValue>(*this).customCSSText();
-    case GridAutoRepeatClass:
-        return downcast<CSSGridAutoRepeatValue>(*this).customCSSText();
-    case GridIntegerRepeatClass:
-        return downcast<CSSGridIntegerRepeatValue>(*this).customCSSText();
-    case GridLineNamesClass:
-        return downcast<CSSGridLineNamesValue>(*this).customCSSText();
-    case SubgridClass:
-        return downcast<CSSSubgridValue>(*this).customCSSText();
-    case GridTemplateAreasClass:
-        return downcast<CSSGridTemplateAreasValue>(*this).customCSSText();
-    case PrimitiveClass:
-        return downcast<CSSPrimitiveValue>(*this).customCSSText(document);
-    case ReflectClass:
-        return downcast<CSSReflectValue>(*this).customCSSText();
-    case ShadowClass:
-        return downcast<CSSShadowValue>(*this).customCSSText();
-    case CubicBezierTimingFunctionClass:
-        return downcast<CSSCubicBezierTimingFunctionValue>(*this).customCSSText();
-    case StepsTimingFunctionClass:
-        return downcast<CSSStepsTimingFunctionValue>(*this).customCSSText();
-    case SpringTimingFunctionClass:
-        return downcast<CSSSpringTimingFunctionValue>(*this).customCSSText();
-    case UnicodeRangeClass:
-        return downcast<CSSUnicodeRangeValue>(*this).customCSSText();
-    case ValueListClass:
-        return downcast<CSSValueList>(*this).customCSSText(document);
-    case ValuePairClass:
-        return downcast<CSSValuePair>(*this).customCSSText();
-    case LineBoxContainClass:
-        return downcast<CSSLineBoxContainValue>(*this).customCSSText();
-    case CalculationClass:
-        return downcast<CSSCalcValue>(*this).customCSSText();
-    case ImageSetClass:
-        return downcast<CSSImageSetValue>(*this).customCSSText();
-    case CSSContentDistributionClass:
-        return downcast<CSSContentDistributionValue>(*this).customCSSText();
-    case CustomPropertyClass:
-        return downcast<CSSCustomPropertyValue>(*this).customCSSText();
-    case VariableReferenceClass:
-        return downcast<CSSVariableReferenceValue>(*this).customCSSText();
-    case PendingSubstitutionValueClass:
-        return downcast<CSSPendingSubstitutionValue>(*this).customCSSText();
-    case OffsetRotateClass:
-        return downcast<CSSOffsetRotateValue>(*this).customCSSText();
-    case RayClass:
-        return downcast<CSSRayValue>(*this).customCSSText();
-    case FontStyleClass:
-        return downcast<CSSFontStyleValue>(*this).customCSSText();
-    case FontStyleRangeClass:
-        return downcast<CSSFontStyleRangeValue>(*this).customCSSText();
-    }
-
-    ASSERT_NOT_REACHED();
-    return String();
+    return visitDerived([](auto& value) {
```

Comment:
If you wanted to maintain the existing behavior of passing a document to some of the derived classes but still keep the nice new visitor, you could implement this like:

```c++
String CSSValue::cssText(Document* document) const
{
    return visitDerived(WTF::makeVisitor(
        [](CSSFunctionValue& value) {
             return value.customCSSText(document);
        },
        [](CSSPrimitiveValue& value) {
             return value.customCSSText(document);
        },
        [](CSSValueList& value) {
             return value.customCSSText(document);
        },
        [](auto& value) {
             return value.customCSSText();
        }
    ));
}
```

(untest code, but a pattern I have used elsewhere).

Reply:
I actually started like that, and ran into all sorts of interesting issues given that some of the classes are derived from other of the classes. I could’ve made it work, but this is much nicer.